gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+With the procedures to construct, compress, and query the sets explained, we'll
+now turn to the modifications to Bitcoin's p2p protocol required to support
+this new operating mode.
+
+
+==== Peer to Peer Service Bit ====
+
+To start, we reserve a currently unutilized service bit. This is required as
+light clients will need to ''preferentially'' peer to full-nodes that support
+the features outlined in this BIP.
+
+The 6th service bit will now be dedicated to signaling support for the
+features described within this BIP: 
+* <code>CFNodeCF = 1 << 6</code>
+
+==== Filter Types ====
+
+As this framework for client-side chain filtering is meant to be generic, in
+this document we define two ''filter types''. A filter type denotes both the
+construction/querying for a filter as well as the contents of the filter.
+
+At the time of writing of this BIP, two filter types are defined:
+* Normal (<code>0x00</code>)
+* Extended (<code>0x01</code>)
+
+A <code>Normal</code> filter is intended to contain all the items that a light client
+needs to sync a basic Bitcoin wallet. In order to facilitate this use-case, for
+each transaction, normal filters contain:
+* The outpoints of each input within a transaction.
+* The data-pushes contained within the public key script of each output within the transaction.
+* The <code>txid</code> of the transaction itself.
+
+An <code>Extended</code> filter contains extra data that is meant to facilitate the
+adoption of more advanced smart contracting applications by this BIP. For each
+transaction found in a block, an <code>Extended</code> filter contains:
+* Each item within the witness stack of an input (if the input has a witness).
+* Each data push of the signature script of an input.
+
+Notably, this construction does not currently interpret P2SH scripts or witness
+scripts to extract data pushes from them; however, future filter types may be
+designed to do so.
+
+==== Filter Construction ====
+
+In order to ensure that filters are deterministically generated, we will use
+the ''first'' <code>16-bytes</code> of the <code>block hash</code> of a Bitcoin
+block as the key to our <code>siphash</code> function. Full-nodes that support
+this BIP are to treat the set of filters as an additional index of the
+blockchain. Once a new block arrives, both filter types should be constructed,
+and stored on disk. Full-nodes that update to support this BIP once already
+synced, should upon start-up, re-index the chain, constructing filters for each
+block from genesis to current chain tip.
+
+Given a Bitcoin block, a full-node constructs a <code>Normal</code> compact
+filter as follows:
+<pre>
+construct_normal_gcs_filter(block, fp) -> []byte:
+    let siphash_key = block.hash()[:16]
+
+    let P = 1 << fp
+
+    let raw_items = []
+    for tx in block.transactions:
+        let txid = tx.hash()
+        raw_items.append(txid)
+
+        for output in tx.outputs:
+            let output_bytes = extract_push_datas(output.script)
```

Comment:
extract_push_datas is never well-defined - only referenced in a general sense.

Reply:
Added a new section explaining what's meant by `extract_push_datas`. I've left a level of abstraction though, as otherwise I'd need to start specifying Script itself...so it currently assumes a certain level of familiarity with Bitcoin (as do all BIPs).